Collisions

Keywarn opened the pull request:
Now implemented pushing other players out the way. Currently, each car calculates locally who should lose the collision and if they believe they have lost they apply forces based on where the other car is.

- Haven't actually tested this over the network and hoping it can be tested at some point this week, just wanted to create the PR so you can see it and it is ready for Friday once tested.

- The amount of 'pushback' can be changed easily using the `collisionAmplifier` variable in the `VehicleDriver`class. I have left it at a reasonable number but this can only really be decided on after some testing.

Resolves #165 
Resolves #164 
Resolves #166

Commit:
Changed car prefab to correct layer

Commit:
Checking for collisions only with other Vehicles/things in the layer mask

Commit:
Now checking to see who's momentum is bigger

Commit:
Add force if losing the collision

Commit:
Made it so cars only collide after a 'cooldown'

Commit:
Better multiplier

Keywarn committed:
Merge branch 'master' into collisions

liamdalg commented on ProjectFiles/Assets/Scripts/Game/Core/Driving/VehicleDriver.cs:
```diff
@@ -79,5 +84,22 @@ public class VehicleDriver : MonoBehaviour {
+                Rigidbody otherBody = other.gameObject.GetComponent<Rigidbody>();
+
+                float myMomentum    = Vector3.Magnitude(body.velocity)      * body.mass;
+                float otherMomentum = Vector3.Magnitude(otherBody.velocity) * otherBody.mass;
```
I feel like this won't work 100% of the time. I think we should resolve the velocity in the direction _between_ the two colliding cars, and do `mass * velocity`, right now it's doing `mass * speed`.

Otherwise, imagine if a car `c1` is travelling in a straight line really fast, and another car `c2` collides perpendicular to it into its side. In the current state, the faster car `c1` will win, even though it's not colliding directly with `c2` no?

```
----------------------------C1------------->
                            ^
                            |
                            |
                            C2

```

(Beautiful diagram)

Keywarn replied:
Yeah, you're right. My code would mean that c1 would win out, so long as c2 wasn't much heavier (depending on the speed). I guess it's down to a gameplay perspective since this isn't very realistic anyway. From a gameplay point of view, is it better for c2 to push c1? I think you're probably right and I'm happy to change it either way.

Just so I understand what you mean by velocity between the cars, you're saying the velocity along the vector I have called `colDir`?

liamdalg replied:
Yeah I think so. I was also thinking, should we maybe do if the momentum difference is within some threshold, they both get sent flying? So then we can have head on collisions be wack

Keywarn replied:
Yeah that’s probably a good idea, I also think currently, head on collision barely move anyone

liamdalg reviewed (dismissed):
Looks good, but check momentum unless I'm wrong

Commit:
Requested changes

- Minimum threshold for collisions to occur
- If within 3x the threshold then both cars are affected
- Who wins is now calculated along the collision direction

Commit:
Changed threshold to be based on the car's mass

Keywarn commented:
Hopefully, a bit more to your taste @liamdalg, should make the collisions much better and added a threshold based on car mass.

I've re-requested review but won't merge this stuff until it gets tested.

ChrisGora commented:
Me and Riley will test now. Fingers crossed 👍🏼

RileyEv committed:
Merge branch 'master' into collisions

RileyEv committed:
build fixes

RileyEv commented on ProjectFiles/Assets/Scripts/Game/Core/Driving/VehicleDriver.cs:
```diff
@@ -1,25 +1,30 @@
         private void Start() {
             body              = gameObject.GetComponent<Rigidbody>();
             body.centerOfMass = transform.Find("centreOfMass").transform.localPosition;
+            collisionThresh = body.mass / 2;
```
Would be better if we removed the threshold and had a curve
